okhttp/src/test/java/okhttp3/CallHandshakeTest.kt

```diff
@@ -0,0 +1,197 @@
+import okhttp3.CipherSuite.Companion.TLS_AES_128_GCM_SHA256
+import okhttp3.CipherSuite.Companion.TLS_AES_256_GCM_SHA384
+import okhttp3.CipherSuite.Companion.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
+import okhttp3.CipherSuite.Companion.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+import okhttp3.testing.PlatformRule
+import okhttp3.tls.internal.TlsUtil.localhost
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+import javax.net.ssl.SSLSocket
+
+class CallHandshakeTest {
+  private lateinit var client: OkHttpClient
+  private lateinit var server: MockWebServer
+  val handshakeCertificates = localhost()
+
+  @RegisterExtension
+  @JvmField
+  val clientTestRule = OkHttpClientTestRule()
+
+  @RegisterExtension
+  @JvmField
+  var platform = PlatformRule()
+
+  private lateinit var handshakeEnabledCipherSuites: List<String>
+  private lateinit var defaultEnabledCipherSuites: List<String>
+  private lateinit var defaultSupportedCipherSuites: List<String>
+
+  val expectedModernTls12CipherSuites = listOf(TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384)
+  val expectedModernTls13CipherSuites = listOf(TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384)
+
+  @BeforeEach
+  fun setup(server: MockWebServer) {
+    this.server = server
+
+    server.enqueue(MockResponse().setResponseCode(200))
+
+    client = clientTestRule.newClientBuilder()
+      .sslSocketFactory(
+        handshakeCertificates.sslSocketFactory(), handshakeCertificates.trustManager)
+      .hostnameVerifier(RecordingHostnameVerifier())
+      .build()
+    server.useHttps(handshakeCertificates.sslSocketFactory(), false)
+
+    defaultEnabledCipherSuites = handshakeCertificates.sslSocketFactory().defaultCipherSuites.toList()
+    defaultSupportedCipherSuites = handshakeCertificates.sslSocketFactory().supportedCipherSuites.toList()
+  }
+
+  @Test
+  fun testDefaultHandshakeCipherSuiteOrderingTls12Restricted() {
+    val client = makeClient(ConnectionSpec.RESTRICTED_TLS, TlsVersion.TLS_1_2)
+
+    val handshake = makeRequest(client)
+
+    assertThat(handshake.cipherSuite).isIn(expectedModernTls12CipherSuites)
+
+    // Probably something like
+    // TLS_AES_128_GCM_SHA256
+    // TLS_AES_256_GCM_SHA384
+    // TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+    // TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+    // TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+    // TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+    assertThat(handshakeEnabledCipherSuites).containsExactlyElementsOf(expectedConnectionCipherSuites(client))
+  }
+
+  @Test
+  fun testDefaultHandshakeCipherSuiteOrderingTls12Modern() {
+    val client = makeClient(ConnectionSpec.MODERN_TLS, TlsVersion.TLS_1_2)
+
+    val handshake = makeRequest(client)
+
+    assertThat(handshake.cipherSuite).isIn(expectedModernTls12CipherSuites)
+
+    // Probably something like
```

Comment:
According to [the doc](https://docs.google.com/spreadsheets/d/1C3FdZSlCBq_-qrVwG1KDIzNIB3Hyg_rKAcgmSzOsHyQ/edit#gid=0) it should be this:

TLS_AES_128_GCM_SHA256
TLS_CHACHA20_POLY1305_SHA256
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
TLS_RSA_WITH_AES_128_GCM_SHA256
TLS_RSA_WITH_AES_256_GCM_SHA384
TLS_RSA_WITH_AES_128_CBC_SHA
TLS_RSA_WITH_AES_256_CBC_SHA
SSL_RSA_WITH_3DES_EDE_CBC_SHA